Handle errors properly

We should not  always just panic

For example, when misspelling the `enabled` property value as "ture" instead of "true", I get a stack trace:

```
panic: yaml: unmarshal errors:
  line 209: cannot unmarshal !!str `ture` into bool

goroutine 1 [running]:
main.check(...)
	/home/ldh/.local/lib/go/src/github.com/binaryplease/base16-universal-manager/main.go:144
main.NewConfig(0xc00001e300, 0x36, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/ldh/.local/lib/go/src/github.com/binaryplease/base16-universal-manager/config.go:48 +0x3f7
main.main()
	/home/ldh/.local/lib/go/src/github.com/binaryplease/base16-universal-manager/main.go:35 +0x78
```

--- a/colorscheme.go
+++ b/colorscheme.go
@@ -1,12 +1,12 @@
 package main
 
 import (
+	"errors"
 	"fmt"
-	"io/ioutil"
 	"os"
+	"path"
 	"strconv"
 	"strings"
-	"path"
 
 	"gopkg.in/yaml.v2"
 )
@@ -35,11 +35,11 @@ type Base16Colorscheme struct {
 	RepoURL string
 
 	RawBaseURL string
-	FileName string
+	FileName   string
 }
 
-func (s Base16Colorscheme) MustacheContext(ext string) map[string]interface{} {
-	var bases = map[string]string{
+func (s Base16Colorscheme) MustacheContext(ext string) (map[string]interface{}, error) {
+	bases := map[string]string{
 		"00": s.Color00,
 		"01": s.Color01,
 		"02": s.Color02,
@@ -59,26 +59,36 @@ func (s Base16Colorscheme) MustacheContext(ext string) map[string]interface{} {
 	}
 	slug := strings.Replace(strings.ToLower(s.FileName), " ", "-", -1)
 	ret := map[string]interface{}{
-		"scheme-name":   s.Name,
-		"scheme-author": s.Author,
-		"scheme-slug": slug,
+		"scheme-name":             s.Name,
+		"scheme-author":           s.Author,
+		"scheme-slug":             slug,
 		"scheme-slug-underscored": strings.Replace(slug, "-", "_", -1),
 	}
 
 	for base, color := range bases {
 		baseKey := "base" + base
 
-		//TODO
+		if len(color) != 6 {
+			return nil, fmt.Errorf("color %q for base %s has incorrect length, must be 6 hex digits", color, base)
+		}
+
+		// TODO
 		rVal := color[:2]
 		gVal := color[2:4]
 		bVal := color[4:6]
 
 		rValf, err := strconv.ParseUint(rVal, 16, 32)
-		check(err)
+		if err != nil {
+			return nil, fmt.Errorf("bad hex red component for base %s: %q", base, rVal)
+		}
 		gValf, err := strconv.ParseUint(gVal, 16, 32)
-		check(err)
+		if err != nil {
+			return nil, fmt.Errorf("bad hex green component for base %s: %q", base, gVal)
+		}
 		bValf, err := strconv.ParseUint(bVal, 16, 32)
-		check(err)
+		if err != nil {
+			return nil, fmt.Errorf("bad hex blue component for base %s: %q", base, bVal)
+		}
 
 		ret[baseKey+"-hex"] = rVal + gVal + bVal
 
@@ -96,98 +106,103 @@ func (s Base16Colorscheme) MustacheContext(ext string) map[string]interface{} {
 
 	}
 
-	return ret
+	return ret, nil
 }
 
 func (l *Base16ColorschemeList) GetBase16Colorscheme(name string) (Base16Colorscheme, error) {
-
 	if len(name) == 0 {
-		panic("Colorscheme name was empty")
+		return Base16Colorscheme{}, errors.New("colorscheme name was empty")
 	}
 
 	schemePath := path.Join(appConf.SchemesCachePath, name)
-
 
 	parts := strings.Split(l.colorschemes[name], "/")
 	yamlURL := strings.Join([]string{"https://raw.githubusercontent.com", parts[3], parts[4], parts[6], parts[7]}, "/")
 	// Create local schemes file, if not present
 	if _, err := os.Stat(schemePath); os.IsNotExist(err) {
 
-		
 		fmt.Println("downloading theme from: ", yamlURL)
 
 		schemeData, err := DownloadFileToString(yamlURL)
-		check(err)
-		saveFile, err := os.Create(schemePath)
-		//TODO delete old file?
-		defer saveFile.Close()
-		check(err)
-		saveFile.Write([]byte(schemeData))
-		saveFile.Close()
+		if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("downloading %s: %w", yamlURL, err)
+		}
+		err = os.WriteFile(schemePath, []byte(schemeData), os.ModePerm)
+		if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("saving scheme to file: %w", err)
+		}
 	}
 
-	colorscheme, err := ioutil.ReadFile(schemePath)
-	check(err)
+	colorscheme, err := os.ReadFile(schemePath)
+	if err != nil {
+		return Base16Colorscheme{}, fmt.Errorf("reading scheme from file %s: %w", schemePath, err)
+	}
 
-	scheme := NewBase16Colorscheme(string(colorscheme))
+	scheme, err := NewBase16Colorscheme(string(colorscheme))
+	if err != nil {
+		return Base16Colorscheme{}, fmt.Errorf("creating scheme from file %s: %w", schemePath, err)
+	}
 	scheme.RawBaseURL = yamlURL
 	scheme.FileName = parts[7][:len(parts[7])-5]
 	return scheme, err
-
 }
 
 func (l *Base16ColorschemeList) GetBase16ColorschemeFile(name string) (Base16Colorscheme, error) {
-
 	if len(name) == 0 {
-		panic("Colorscheme path was empty")
+		return Base16Colorscheme{}, fmt.Errorf("colorscheme path was empty")
 	}
 
-	_, err := os.Stat(name)
-	// no error is tolerable for local files
-	check(err)
+	colorscheme, err := os.ReadFile(name)
+	if err != nil {
+		return Base16Colorscheme{}, fmt.Errorf("reading scheme from file %s: %w", name, err)
+	}
 
-	colorscheme, err := ioutil.ReadFile(name)
-	check(err)
-
-	scheme := NewBase16Colorscheme(string(colorscheme))
+	scheme, err := NewBase16Colorscheme(string(colorscheme))
+	if err != nil {
+		return Base16Colorscheme{}, fmt.Errorf("creating scheme from file %s: %w", name, err)
+	}
 	_, fileName := path.Split(name)
 	scheme.FileName = fileName[:len(fileName)-5]
 
 	return scheme, err
-
 }
 
-func NewBase16Colorscheme(yamlData string) Base16Colorscheme {
+func NewBase16Colorscheme(yamlData string) (Base16Colorscheme, error) {
 	var scheme Base16Colorscheme
 
 	err := yaml.Unmarshal([]byte(yamlData), &scheme)
-	check(err)
-	return scheme
+	if err != nil {
+		return Base16Colorscheme{}, fmt.Errorf("unmarshalling scheme from YAML: %w", err)
+	}
+	return scheme, nil
 }
 
-func LoadBase16ColorschemeList() Base16ColorschemeList {
-	colorschemes := LoadStringMap(appConf.SchemesListFile)
-	return Base16ColorschemeList{colorschemes}
+func LoadBase16ColorschemeList() (Base16ColorschemeList, error) {
+	colorschemes, err := LoadStringMap(appConf.SchemesListFile)
+	return Base16ColorschemeList{colorschemes}, err
 }
 
-func SaveBase16ColorschemeList(l Base16ColorschemeList) {
-	SaveStringMap(l.colorschemes, appConf.SchemesListFile)
+func SaveBase16ColorschemeList(l Base16ColorschemeList) error {
+	return SaveStringMap(l.colorschemes, appConf.SchemesListFile)
 }
 
 type Base16ColorschemeList struct {
 	colorschemes map[string]string
 }
 
-func (l *Base16ColorschemeList) UpdateSchemes() {
-
-	//Get all repos from master source
+func (l *Base16ColorschemeList) UpdateSchemes() error {
+	// Get all repos from master source
 	schemeRepos := make(map[string]string)
 
 	schemesYAML, err := DownloadFileToString(appConf.SchemesMasterURL)
-	check(err)
+	if err != nil {
+		return fmt.Errorf("downloading scheme from %s: %w", appConf.SchemesMasterURL, err)
+	}
 
 	err = yaml.Unmarshal([]byte(schemesYAML), &schemeRepos)
-	check(err)
+	if err != nil {
+		return fmt.Errorf("unmarshalling YAML: %w", err)
+	}
 
 	fmt.Println("Found colorscheme repos: ", len(schemeRepos))
 
@@ -205,32 +220,49 @@ func (l *Base16ColorschemeList) UpdateSchemes() {
 	}
 
 	fmt.Println("Found colorschemes: ", len(l.colorschemes))
-	SaveBase16ColorschemeList(Base16ColorschemeList{l.colorschemes})
+	if err := SaveBase16ColorschemeList(Base16ColorschemeList{l.colorschemes}); err != nil {
+		return fmt.Errorf("saving colorscheme list: %w", err)
+	}
+
+	return nil
 }
 
-func (c *Base16ColorschemeList) Find(input string) Base16Colorscheme {
+func (c *Base16ColorschemeList) Find(input string) (Base16Colorscheme, error) {
 	if strings.Contains(input, "/") {
 		// a local scheme path, not a name
 		colorschemeName := input
 		scheme, err := c.GetBase16ColorschemeFile(colorschemeName)
-		check(err)
-		return scheme
+		if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("getting colorscheme file for %s: %w", colorschemeName, err)
+		}
+		return scheme, nil
 	} else {
 		// a name, look it up
-		if _, err := os.Stat(appConf.SchemesListFile); os.IsNotExist(err) {
-			check(err)
+		_, err := os.Stat(appConf.SchemesListFile)
+		if os.IsNotExist(err) {
 			fmt.Println("Colorschemes list not found, pulling new one...")
-			c.UpdateSchemes()
+			if err := c.UpdateSchemes(); err != nil {
+				return Base16Colorscheme{}, fmt.Errorf("updating schemes: %w", err)
+			}
+		} else if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("checking for existing colorschems list: %w", err)
 		}
 
 		if len(c.colorschemes) == 0 {
 			fmt.Println("No templates in list, pulling new one... ")
-			c.UpdateSchemes()
+			if err := c.UpdateSchemes(); err != nil {
+				return Base16Colorscheme{}, fmt.Errorf("updating schemes: %w", err)
+			}
 		}
 
-		colorschemeName := FindMatchInMap(c.colorschemes, input)
+		colorschemeName, err := FindMatchInMap(c.colorschemes, input)
+		if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("finding colorscheme in list: %w", err)
+		}
 		scheme, err := c.GetBase16Colorscheme(colorschemeName)
-		check(err)
-		return scheme
+		if err != nil {
+			return Base16Colorscheme{}, fmt.Errorf("getting colorscheme %s: %w", colorschemeName, err)
+		}
+		return scheme, nil
 	}
 }

--- a/colorscheme_test.go
+++ b/colorscheme_test.go
@@ -61,7 +61,11 @@ func TestBase16Colorscheme_MustacheContext(t *testing.T) {
 				RawBaseURL: tt.fields.RawBaseURL,
 				FileName:   tt.fields.FileName,
 			}
-			if got := s.MustacheContext(".config"); !reflect.DeepEqual(got, tt.want) {
+			got, err := s.MustacheContext(".config")
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("Base16Colorscheme.MustacheContext() = %v, want %v", got, tt.want)
 			}
 		})
@@ -152,7 +156,11 @@ base0F: "c33ff3"`
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := NewBase16Colorscheme(tt.yamlData); !reflect.DeepEqual(got, tt.want) {
+			got, err := NewBase16Colorscheme(tt.yamlData)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("NewBase16Colorscheme() = %v, want %v", got, tt.want)
 			}
 		})
@@ -168,7 +176,11 @@ func TestLoadBase16ColorschemeList(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := LoadBase16ColorschemeList(); !reflect.DeepEqual(got, tt.want) {
+			got, err := LoadBase16ColorschemeList()
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("LoadBase16ColorschemeList() = %v, want %v", got, tt.want)
 			}
 		})
@@ -187,7 +199,9 @@ func TestSaveBase16ColorschemeList(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			SaveBase16ColorschemeList(tt.args.l)
+			if err := SaveBase16ColorschemeList(tt.args.l); err != nil {
+				t.Error(err)
+			}
 		})
 	}
 }
@@ -207,7 +221,9 @@ func TestBase16ColorschemeList_UpdateSchemes(t *testing.T) {
 			l := &Base16ColorschemeList{
 				colorschemes: tt.fields.colorschemes,
 			}
-			l.UpdateSchemes()
+			if err := l.UpdateSchemes(); err != nil {
+				t.Error(err)
+			}
 		})
 	}
 }
@@ -232,7 +248,11 @@ func TestBase16ColorschemeList_Find(t *testing.T) {
 			c := &Base16ColorschemeList{
 				colorschemes: tt.fields.colorschemes,
 			}
-			if got := c.Find(tt.args.input); !reflect.DeepEqual(got, tt.want) {
+			got, err := c.Find(tt.args.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("Base16ColorschemeList.Find() = %v, want %v", got, tt.want)
 			}
 		})

--- a/config.go
+++ b/config.go
@@ -1,16 +1,18 @@
 package main
 
 import (
+	"errors"
 	"fmt"
-	"io/ioutil"
 	"os"
 	"path/filepath"
 
 	"gopkg.in/yaml.v2"
 )
 
-var defaultSchemesMasterURL = "https://raw.githubusercontent.com/chriskempson/base16-schemes-source/master/list.yaml"
-var defaultTemplatesMasterURL = "https://raw.githubusercontent.com/chriskempson/base16-templates-source/master/list.yaml"
+var (
+	defaultSchemesMasterURL   = "https://raw.githubusercontent.com/chriskempson/base16-schemes-source/master/list.yaml"
+	defaultTemplatesMasterURL = "https://raw.githubusercontent.com/chriskempson/base16-templates-source/master/list.yaml"
+)
 
 // SetterConfig is the applicaton's configuration.
 type SetterConfig struct {
@@ -44,20 +46,20 @@ type FileConfig struct {
 }
 
 // NewConfig parses the provided configuration file and returns the app configuration.
-func NewConfig(path string) SetterConfig {
+func NewConfig(path string) (SetterConfig, error) {
 	if path == "" {
-		fmt.Fprintf(os.Stderr, "no config file found\n")
-		os.Exit(1)
+		return SetterConfig{}, errors.New("no config file specified")
 	}
 
 	var conf SetterConfig
-	file, err := ioutil.ReadFile(path)
+	file, err := os.ReadFile(path)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "error: %v\n", err)
-		os.Exit(1)
+		return SetterConfig{}, fmt.Errorf("reading config file: %w", err)
 	}
 	err = yaml.Unmarshal(file, &conf)
-	check(err)
+	if err != nil {
+		return SetterConfig{}, fmt.Errorf("parsing YAML from config file %s: %w", path, err)
+	}
 
 	if conf.SchemesMasterURL == "" {
 		conf.SchemesMasterURL = defaultSchemesMasterURL
@@ -71,7 +73,7 @@ func NewConfig(path string) SetterConfig {
 	conf.TemplatesCachePath = filepath.Join(xdgDirs.CacheHome(), "templates")
 	conf.TemplatesListFile = filepath.Join(xdgDirs.CacheHome(), "templateslist.yaml")
 
-	return conf
+	return conf, nil
 }
 
 // Show prints the app configuration

--- a/config_test.go
+++ b/config_test.go
@@ -1,7 +1,6 @@
 package main
 
 import (
-	"io/ioutil"
 	"os"
 	"reflect"
 	"testing"
@@ -20,7 +19,11 @@ func TestNewConfig(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := NewConfig(tt.args.path); !reflect.DeepEqual(got, tt.want) {
+			got, err := NewConfig(tt.args.path)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("NewConfig() = %v, want %v", got, tt.want)
 			}
 		})
@@ -67,7 +70,7 @@ func TestSetterConfig_Show(t *testing.T) {
 
 func TestDefaultMasterURLs(t *testing.T) {
 	// Create a temporary, empty config file
-	tmpFile, err := ioutil.TempFile(os.TempDir(), "base16-universal-manager-")
+	tmpFile, err := os.CreateTemp(os.TempDir(), "base16-universal-manager-")
 	if err != nil {
 		t.Fatalf("Cannot create temporary file\n")
 	}
@@ -76,7 +79,10 @@ func TestDefaultMasterURLs(t *testing.T) {
 	}
 	defer os.Remove(tmpFile.Name())
 
-	config := NewConfig(tmpFile.Name())
+	config, err := NewConfig(tmpFile.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
 	if config.SchemesMasterURL == "" {
 		t.Fatalf("SchemesMasterURL should default to %s\n", defaultSchemesMasterURL)
 	}

--- a/helpers.go
+++ b/helpers.go
@@ -125,18 +125,30 @@ func FindMatchInMap(choices map[string]string, input string) (string, error) {
 	if len(choices) == 0 {
 		return "", errors.New("cannot select from empty choices")
 	}
-	var match string
-	distance := 1000
+	var matches []string
+	var distance int
 
 	for k := range choices {
 		tempDistance := levenshtein.ComputeDistance(input, k)
-		if tempDistance < distance {
-			match = k
+		if tempDistance < distance || len(matches) == 0 {
 			distance = tempDistance
+			matches = nil
+			matches = append(matches, k)
+		} else if tempDistance == distance {
+			matches = append(matches, k)
 		}
 	}
 
-	return match, nil
+	if len(matches) != 1 {
+		return "", fmt.Errorf(
+			"found %d matches to %q with same similarity score, consider specifying a closer match to a candidate - candidates are: %v",
+			len(matches),
+			input,
+			strings.Join(matches, ", "),
+		)
+	}
+
+	return matches[0], nil
 }
 
 func exe_cmd(cmd string) {

--- a/helpers.go
+++ b/helpers.go
@@ -1,12 +1,12 @@
 package main
 
 import (
-	"encoding/json"
-
 	"bufio"
 	"bytes"
+	"encoding/json"
+	"errors"
 	"fmt"
-	"io/ioutil"
+	"io"
 	"log"
 	"net/http"
 	"os"
@@ -28,7 +28,7 @@ func DownloadFileToString(url string) (string, error) {
 	}
 
 	if appConf.GithubToken != "" {
-		req.Header.Add("Authorization", "token " + appConf.GithubToken)
+		req.Header.Add("Authorization", "token "+appConf.GithubToken)
 	}
 
 	resp, err := client.Do(req)
@@ -37,7 +37,7 @@ func DownloadFileToString(url string) (string, error) {
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode == http.StatusOK {
-		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		bodyBytes, err := io.ReadAll(resp.Body)
 		if err != nil {
 			return "", err
 		}
@@ -70,7 +70,6 @@ type GitHubFilesCollection struct {
 }
 
 func findYAMLinRepo(repoURL string) []GitHubFile {
-
 	parts := strings.Split(repoURL, "/")
 	ApiUrl := ("https://api.github.com/repos/" + parts[3] + "/" + parts[4] + "/contents/")
 	// fmt.Println("generated api URL: ", ApiUrl)
@@ -98,34 +97,33 @@ func findYAMLinRepo(repoURL string) []GitHubFile {
 	return colorSchemes
 }
 
-func LoadStringMap(path string) map[string]string {
-
-	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0600)
-	check(err)
-	yamlFile, err := ioutil.ReadAll(f)
-	check(err)
+func LoadStringMap(path string) (map[string]string, error) {
+	yamlFile, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("reading from file %s: %w", path, err)
+	}
 	data := make(map[string]string)
-	err = yaml.Unmarshal(yamlFile, data)
-	check(err)
-	return data
+	if err := yaml.Unmarshal(yamlFile, data); err != nil {
+		return nil, fmt.Errorf("parsing YAML from file %s: %w", path, err)
+	}
+	return data, nil
 }
 
-func SaveStringMap(data map[string]string, path string) {
-
+func SaveStringMap(data map[string]string, path string) error {
 	yamlData, err := yaml.Marshal(data)
-	check(err)
-	saveFile, err := os.Create(path)
-	defer saveFile.Close()
-	saveFile.Write(yamlData)
-	saveFile.Close()
-	fmt.Println("wrote to: ", saveFile.Name())
+	if err != nil {
+		return fmt.Errorf("marshaling to YAML: %w", err)
+	}
+	if err := os.WriteFile(path, yamlData, os.ModePerm); err != nil {
+		return fmt.Errorf("writing YAML: %w", err)
+	}
+	fmt.Println("wrote to: ", path)
+	return nil
 }
 
-func FindMatchInMap(choices map[string]string, input string) string {
-
+func FindMatchInMap(choices map[string]string, input string) (string, error) {
 	if len(choices) == 0 {
-		panic("cannot select from empty choices")
-
+		return "", errors.New("cannot select from empty choices")
 	}
 	var match string
 	distance := 1000
@@ -138,17 +136,16 @@ func FindMatchInMap(choices map[string]string, input string) string {
 		}
 	}
 
-	return match
+	return match, nil
 }
 
 func exe_cmd(cmd string) {
-
 	if len(cmd) == 0 {
 		return
 	}
 	parts := strings.Fields(cmd)
 	head := parts[0]
-	parts = parts[1:len(parts)]
+	parts = parts[1:]
 
 	out, err := exec.Command(head, parts...).Output()
 

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -66,7 +66,11 @@ func TestLoadStringMap(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := LoadStringMap(tt.args.path); !reflect.DeepEqual(got, tt.want) {
+			got, err := LoadStringMap(tt.args.path)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("LoadStringMap() = %v, want %v", got, tt.want)
 			}
 		})
@@ -86,7 +90,9 @@ func TestSaveStringMap(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			SaveStringMap(tt.args.data, tt.args.path)
+			if err := SaveStringMap(tt.args.data, tt.args.path); err != nil {
+				t.Error(err)
+			}
 		})
 	}
 }
@@ -105,7 +111,11 @@ func TestFindMatchInMap(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := FindMatchInMap(tt.args.choices, tt.args.input); got != tt.want {
+			got, err := FindMatchInMap(tt.args.choices, tt.args.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if got != tt.want {
 				t.Errorf("FindMatchInMap() = %v, want %v", got, tt.want)
 			}
 		})
@@ -176,8 +186,7 @@ func TestWriteFile(t *testing.T) {
 }
 
 func TestReplaceMultiline(t *testing.T) {
-	type args struct {
-	}
+	type args struct{}
 	tests := []struct {
 		name        string
 		path        string

--- a/main.go
+++ b/main.go
@@ -32,11 +32,21 @@ var (
 var appConf SetterConfig
 
 func main() {
-	//Parse Flags
+	if err := run(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+	}
+}
+
+func run() error {
+	// Parse Flags
 	kingpin.Version("1.0.0")
 	kingpin.Parse()
 
-	appConf = NewConfig(*configFileFlag)
+	var err error
+	appConf, err = NewConfig(*configFileFlag)
+	if err != nil {
+		return err
+	}
 
 	if *printConfigFlag {
 		appConf.Show()
@@ -79,21 +89,37 @@ func main() {
 	os.MkdirAll(appConf.SchemesCachePath, os.ModePerm)
 	os.MkdirAll(appConf.TemplatesCachePath, os.ModePerm)
 
-	schemeList := LoadBase16ColorschemeList()
-	templateList := LoadBase16TemplateList()
+	schemeList, err := LoadBase16ColorschemeList()
+	if err != nil {
+		return fmt.Errorf("loading colorscheme list: %w", err)
+	}
+	templateList, err := LoadBase16TemplateList()
+	if err != nil {
+		return fmt.Errorf("loading template list: %w", err)
+	}
 
 	if *updateFlag {
-		schemeList.UpdateSchemes()
-		templateList.UpdateTemplates()
+		if err := schemeList.UpdateSchemes(); err != nil {
+			return fmt.Errorf("updating schemes: %w", err)
+		}
+		if err := templateList.UpdateTemplates(); err != nil {
+			return fmt.Errorf("updating templates: %w", err)
+		}
 	}
 
 	var scheme Base16Colorscheme
 	if *schemeFlag == "" {
 		// Scheme from config
-		scheme = schemeList.Find(appConf.Colorscheme)
+		scheme, err = schemeList.Find(appConf.Colorscheme)
+		if err != nil {
+			return fmt.Errorf("finding scheme %s from configuration: %w", appConf.Colorscheme, err)
+		}
 	} else {
 		// Scheme from flag
-		scheme = schemeList.Find(*schemeFlag)
+		scheme, err = schemeList.Find(*schemeFlag)
+		if err != nil {
+			return fmt.Errorf("finding scheme %s from flag: %w", *schemeFlag, err)
+		}
 	}
 	fmt.Println("[CONFIG]: Selected scheme: ", scheme.Name)
 
@@ -103,9 +129,13 @@ func main() {
 			appConfig.Template = app
 		}
 		if appConfig.Enabled {
-			err := Base16Render(templateList.Find(appConfig.Template), scheme, app)
+			tmpl, err := templateList.Find(appConfig.Template)
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "Error when rendering file: %v\n", err)
+				return fmt.Errorf("finding template %s: %w", appConfig.Template, err)
+			}
+			err = Base16Render(tmpl, scheme, app)
+			if err != nil {
+				return fmt.Errorf("rendering file: %v", err)
 			}
 			templateEnabled = true
 		}
@@ -115,6 +145,7 @@ func main() {
 		fmt.Println("No templates enabled")
 	}
 
+	return nil
 }
 
 // Base16Render takes an application-specific template and renders a config file
@@ -127,7 +158,11 @@ func Base16Render(templ Base16Template, scheme Base16Colorscheme, app string) er
 		if err != nil {
 			return fmt.Errorf("could not download template file: %w", err)
 		}
-		renderedFile := mustache.Render(templFileData, scheme.MustacheContext(v.Extension))
+		templContext, err := scheme.MustacheContext(v.Extension)
+		if err != nil {
+			return fmt.Errorf("generating template context: %w", err)
+		}
+		renderedFile := mustache.Render(templFileData, templContext)
 
 		savePath, err := getSavePath(appConf.Applications[app].Files[k].Path, k+v.Extension)
 		if err != nil {
@@ -137,7 +172,7 @@ func Base16Render(templ Base16Template, scheme Base16Colorscheme, app string) er
 			continue
 		}
 
-		//If DryRun is enabled, just print the output location for debugging
+		// If DryRun is enabled, just print the output location for debugging
 		if appConf.DryRun {
 			fmt.Println("    - (dryrun) file would be written to: ", savePath)
 		} else {
@@ -201,11 +236,4 @@ func getSavePath(path, defaultFilename string) (string, error) {
 	}
 
 	return savePath, nil
-}
-
-// TODO proper error handling
-func check(e error) {
-	if e != nil {
-		panic(e)
-	}
 }

--- a/main_test.go
+++ b/main_test.go
@@ -21,7 +21,7 @@ func TestBase16Render(t *testing.T) {
 	type args struct {
 		templ  Base16Template
 		scheme Base16Colorscheme
-		app string
+		app    string
 	}
 	tests := []struct {
 		name string
@@ -32,23 +32,6 @@ func TestBase16Render(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			Base16Render(tt.args.templ, tt.args.scheme, tt.args.app)
-		})
-	}
-}
-
-func Test_check(t *testing.T) {
-	type args struct {
-		e error
-	}
-	tests := []struct {
-		name string
-		args args
-	}{
-		// TODO: Add test cases.
-	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			check(tt.args.e)
 		})
 	}
 }

--- a/template.go
+++ b/template.go
@@ -1,12 +1,13 @@
 package main
 
 import (
+	"errors"
 	"fmt"
-	"gopkg.in/yaml.v2"
-	"io/ioutil"
 	"os"
 	"path"
 	"strings"
+
+	"gopkg.in/yaml.v2"
 )
 
 type Base16TemplateFile struct {
@@ -15,11 +16,10 @@ type Base16TemplateFile struct {
 }
 
 type Base16Template struct {
-
-	//The actual template
+	// The actual template
 	Files map[string]Base16TemplateFile
 
-	//Name (of the application)
+	// Name (of the application)
 	Name string
 
 	RepoURL string
@@ -28,7 +28,6 @@ type Base16Template struct {
 }
 
 func GetRawBaseURL(repoURL string, mainBranch string) (string, error) {
-
 	parts := strings.Split(repoURL, "/")
 	rawBaseURL := ""
 	repoName := parts[3] + "/" + parts[4]
@@ -45,17 +44,18 @@ func GetRawBaseURL(repoURL string, mainBranch string) (string, error) {
 	return rawBaseURL, nil
 }
 
-func (l *Base16TemplateList) GetBase16Template(name string, remoteBranch string) Base16Template {
-
+func (l *Base16TemplateList) GetBase16Template(name string, remoteBranch string) (Base16Template, error) {
 	// yamlURL := "https://raw.githubusercontent.com/" + parts[3] + "/" + parts[4] + "/master/templates/config.yaml"
 	if len(name) == 0 {
-		panic("Template name was empty")
+		return Base16Template{}, errors.New("template name was empty")
 	}
 
 	var newTemplate Base16Template
 	newTemplate.RepoURL = l.templates[name]
 	rawBaseURL, err := GetRawBaseURL(l.templates[name], remoteBranch)
-	check(err)
+	if err != nil {
+		return Base16Template{}, fmt.Errorf("generating template URL: %w", err)
+	}
 	newTemplate.RawBaseURL = rawBaseURL
 	newTemplate.Name = name
 
@@ -63,75 +63,95 @@ func (l *Base16TemplateList) GetBase16Template(name string, remoteBranch string)
 
 	// Create local template file, if not present
 	if _, err := os.Stat(templatePath); os.IsNotExist(err) {
-		templateData, err := DownloadFileToString(newTemplate.RawBaseURL + "templates/config.yaml")
-		check(err)
+		templateURL := newTemplate.RawBaseURL + "templates/config.yaml"
+		templateData, err := DownloadFileToString(templateURL)
+		if err != nil {
+			return Base16Template{}, fmt.Errorf("downloading template from %s: %w", templateURL, err)
+		}
 		saveFile, err := os.Create(templatePath)
-		//TODO delete old file?
+		// TODO delete old file?
 		defer saveFile.Close()
 		saveFile.Write([]byte(templateData))
 		saveFile.Close()
 	}
 
-	template, err := ioutil.ReadFile(templatePath)
-	check(err)
+	template, err := os.ReadFile(templatePath)
+	if err != nil {
+		return Base16Template{}, fmt.Errorf("reading template file: %w", err)
+	}
 
-	//TODO cache actual templates
+	// TODO cache actual templates
 	var files map[string]Base16TemplateFile
 
 	err = yaml.Unmarshal(template, &files)
-	check(err)
+	if err != nil {
+		return Base16Template{}, fmt.Errorf("parsing template file %q: %w", templatePath, err)
+	}
 
 	newTemplate.Files = files
-	return newTemplate
+	return newTemplate, nil
 }
 
 type Base16TemplateList struct {
 	templates map[string]string
 }
 
-func (l *Base16TemplateList) UpdateTemplates() {
-
-	//Get all repos from master source
+func (l *Base16TemplateList) UpdateTemplates() error {
+	// Get all repos from master source
 	var templRepos map[string]string
 
 	templatesYAML, err := DownloadFileToString(appConf.TemplatesMasterURL)
-	check(err)
+	if err != nil {
+		return fmt.Errorf("downloading file from %s: %w", appConf.TemplatesMasterURL, err)
+	}
 
 	err = yaml.Unmarshal([]byte(templatesYAML), &templRepos)
-	check(err)
+	if err != nil {
+		return fmt.Errorf("parsing file downloaded from %s: %w", appConf.TemplatesMasterURL, err)
+	}
 
 	fmt.Println("Found template repos: ", len(templRepos))
 	for k, v := range templRepos {
 		l.templates[k] = v
 	}
 
-	SaveBase16TemplateList(Base16TemplateList{l.templates})
+	if err := SaveBase16TemplateList(Base16TemplateList{l.templates}); err != nil {
+		return fmt.Errorf("saving template list: %w", err)
+	}
 
+	return nil
 }
 
-func LoadBase16TemplateList() Base16TemplateList {
-	colorschemes := LoadStringMap(appConf.TemplatesListFile)
-	return Base16TemplateList{colorschemes}
+func LoadBase16TemplateList() (Base16TemplateList, error) {
+	colorschemes, err := LoadStringMap(appConf.TemplatesListFile)
+	return Base16TemplateList{colorschemes}, err
 }
 
-func SaveBase16TemplateList(l Base16TemplateList) {
-	SaveStringMap(l.templates, appConf.TemplatesListFile)
+func SaveBase16TemplateList(l Base16TemplateList) error {
+	return SaveStringMap(l.templates, appConf.TemplatesListFile)
 }
 
-func (c *Base16TemplateList) Find(input string) Base16Template {
-
+func (c *Base16TemplateList) Find(input string) (Base16Template, error) {
 	if _, err := os.Stat(appConf.TemplatesListFile); os.IsNotExist(err) {
-		check(err)
 		fmt.Println("Templates list not found, pulling new one...")
-		c.UpdateTemplates()
+		if err := c.UpdateTemplates(); err != nil {
+			return Base16Template{}, fmt.Errorf("updating templates: %w", err)
+		}
+	} else if err != nil {
+		return Base16Template{}, fmt.Errorf("checking existance of template list: %w", err)
 	}
 
 	if len(c.templates) == 0 {
 		fmt.Println("No templates in list, pulling new one... ")
-		c.UpdateTemplates()
+		if err := c.UpdateTemplates(); err != nil {
+			return Base16Template{}, fmt.Errorf("updating templates: %w", err)
+		}
 	}
 
-	templateName := FindMatchInMap(c.templates, input)
+	templateName, err := FindMatchInMap(c.templates, input)
+	if err != nil {
+		return Base16Template{}, fmt.Errorf("finding template in list: %w", err)
+	}
 	appConfig := appConf.Applications[input]
 	remoteBranch := "master"
 	if appConfig.DefaultRemoteBranch != "" {

--- a/template_test.go
+++ b/template_test.go
@@ -73,7 +73,11 @@ func TestBase16TemplateList_GetBase16Template(t *testing.T) {
 			l := &Base16TemplateList{
 				templates: tt.fields.templates,
 			}
-			if got := l.GetBase16Template(tt.args.name, "master"); !reflect.DeepEqual(got, tt.want) {
+			got, err := l.GetBase16Template(tt.args.name, "master")
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("Base16TemplateList.GetBase16Template() = %v, want %v", got, tt.want)
 			}
 		})
@@ -109,7 +113,11 @@ func TestLoadBase16TemplateList(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := LoadBase16TemplateList(); !reflect.DeepEqual(got, tt.want) {
+			got, err := LoadBase16TemplateList()
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("LoadBase16TemplateList() = %v, want %v", got, tt.want)
 			}
 		})
@@ -128,7 +136,9 @@ func TestSaveBase16TemplateList(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			SaveBase16TemplateList(tt.args.l)
+			if err := SaveBase16TemplateList(tt.args.l); err != nil {
+				t.Error(err)
+			}
 		})
 	}
 }
@@ -153,7 +163,11 @@ func TestBase16TemplateList_Find(t *testing.T) {
 			c := &Base16TemplateList{
 				templates: tt.fields.templates,
 			}
-			if got := c.Find(tt.args.input); !reflect.DeepEqual(got, tt.want) {
+			got, err := c.Find(tt.args.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("Base16TemplateList.Find() = %v, want %v", got, tt.want)
 			}
 		})

